crates/core/src/context.rs

```diff
@@ -86,6 +89,7 @@ pub struct Context {
     pub(crate) signal_tx: Option<Sender<Signal>>,
     pub(crate) instance_is_alive: Arc<AtomicBool>,
     pub state_dump_logging: bool,
+    pub metric_publisher: Arc<RwLock<dyn MetricPublisher>>,
```

Comment:
It seems that the only reason this needs to be in a mutex is because of the mutable reference to self in `publish`. So can we talk about if we really need that? Any logging/reporting should be as lightweight as possible.